src/aws-cpp-sdk-core/include/aws/core/utils/threading/DefaultExecutor.h

```diff
@@ -31,14 +31,39 @@ namespace Aws
 
                 void WaitUntilStopped() override;
             protected:
+                struct DefaultExecutorTask {
+                  DefaultExecutorTask(std::function<void()>&& task, DefaultExecutor* executor)
+                    : m_task(std::move(task)), m_executor(executor) {
+                    assert(m_task);
+                  }
+
+                  DefaultExecutorTask(const DefaultExecutorTask&) = delete;
+                  DefaultExecutorTask& operator=(const DefaultExecutorTask&) = delete;
+                  DefaultExecutorTask(DefaultExecutorTask&&) = default;
+
+                  void Execute() {
+                    assert(m_task);
+                    m_task();
+                    if (m_executor) {
+                      m_executor->Detach(std::this_thread::get_id());
+                    }
+                    Aws::Delete(this);
```

Comment:
im not sure i love this because a segfault for calling execute twice seems like a bad failure mode for this. could we maybe place a level indirection above the task that prevents it from being called twice ever?

Reply:
I has the same thought... if you think about it more, after the task has been destructed, you cannot trust anything under `this->`, as it is already destroyed.

The indirection here is 
```
using DefaultExecutorTaskPair = std::pair<std::thread, DefaultExecutorTask*>;
Aws::UnorderedMap<std::thread::id, DefaultExecutorTaskPair> m_tasks;
```

I.e. the indirection is thread_id used as a key.
We need this structure to *update* the task *after* it has been launched, so we need to have access to the task context from the Executor, but also we need to allow the Task to self-destroy.
There is another potential issue: Executor cannot access/update any DefaultExecutorTask (this will create a race condition), it can do it if being called from the same thread of the Task.